### ISSUE TITLE
`curator tag -s audio` cannot tag AVI files

### ISSUE THREAD

**Issue by alanorth** [AlexAltea/curator#8]
I noticed that `curator tag -s audio` doesn't seem to be able to tag audio files. Running this multiple times shows that nothing has changed, so the tag never gets set:

```console
$ chrt -b 0 curator tag -s audio -t language --only-macrolanguages --log=DEBUG --max-audio-samples=2 movie.avi
2023-01-29 11:10:03,468 | INFO | Analyzing 1 input media files
2023-01-29 11:10:03,677 | DEBUG | Detecting audio language in stream #1 of media: "movie.avi"
2023-01-29 11:10:05,968 | DEBUG | Sample #00: {'en': '0.4914', 'la': '0.0963', 'ja': '0.0866', 'zh': '0.0579', 'ru': '0.0533'}
2023-01-29 11:10:08,650 | DEBUG | Sample #01: {'en': '0.9920', 'nn': '0.0016', 'tl': '0.0007', 'es': '0.0007', 'ms': '0.0004'}
┌───┬─────────────────────────┬────────┬─────────┬───┬─────────┐
│ # │ Name                    │ Stream │ Old tag │ → │ New tag │
├───┼─────────────────────────┼────────┼─────────┼───┼─────────┤
│ 1 │ movie.avi               │ 1      │ None    │ → │ eng     │
└───┴─────────────────────────┴────────┴─────────┴───┴─────────┘
Continue? [y/N] y
$ chrt -b 0 curator tag -s audio -t language --only-macrolanguages --log=DEBUG --max-audio-samples=2 movie.avi
2023-01-29 11:10:16,224 | INFO | Analyzing 1 input media files
2023-01-29 11:10:16,285 | DEBUG | Detecting audio language in stream #1 of media: "movie.avi"
2023-01-29 11:10:18,244 | DEBUG | Sample #00: {'en': '0.4914', 'la': '0.0963', 'ja': '0.0866', 'zh': '0.0579', 'ru': '0.0533'}
2023-01-29 11:10:19,330 | DEBUG | Sample #01: {'en': '0.9920', 'nn': '0.0016', 'tl': '0.0007', 'es': '0.0007', 'ms': '0.0004'}
┌───┬─────────────────────────┬────────┬─────────┬───┬─────────┐
│ # │ Name                    │ Stream │ Old tag │ → │ New tag │
├───┼─────────────────────────┼────────┼─────────┼───┼─────────┤
│ 1 │ movie.avi               │ 1      │ None    │ → │ eng     │
└───┴─────────────────────────┴────────┴─────────┴───┴─────────┘
Continue? [y/N] y
```

**Comment by AlexAltea** [comment]
This is caused by a shortcoming for RIFF/AVI contianers. There's no per-stream metadata, only container-wide metadata.

Nonetheless `IAS1-9` tags could tag the language for the first 9 audio stream, so that might solve this problem.\
Solution is described here: https://superuser.com/a/1223919

I will implement this tonight when I'm back home, latest tomorrow.

**Comment by alanorth** [comment]
Ah, that's interesting. I suppose it's a corner case for legacy media files, so not a huge priority for me, unless you want to handle such corner cases. From my side, for these old files there, are a few options:

1. Copy streams from AVI container to MKV/MP4 container (brings up [another issue](https://stackoverflow.com/questions/66013483/cant-write-packet-with-unknown-timestamp-av-interleaved-write-frame-invalid) in my testing)
2. Download new copies (these are shitty AVI containers with shitty Xvid video codec anyways)
3. Delete the movies. :)

**Comment by AlexAltea** [comment]
> Copy streams from AVI container to MKV/MP4 container (brings up [another issue](https://stackoverflow.com/questions/66013483/cant-write-packet-with-unknown-timestamp-av-interleaved-write-frame-invalid) in my testing)

I'm doing that precisely. With `curator convert -f mkv` (I haven't documented this tool yet).

It addresses `+genpts` and other common issues when changing media formats: https://github.com/AlexAltea/curator/blob/6d6e211cd2be1bc7f7eb7bc8657944d724be74f9/curator/plans/convert.py#L78-L92

> Download new copies (these are shitty AVI containers with shitty Xvid video codec anyways)

Yes, that's what I am planning to do with my own collection as well! Crawling the internet for new versions and then using `curator merge`. I want to do merge instead of just replace, to preserve dubs from older rips that might not exist in better higher-quality rips in their original language.

Ideally, `curator merge` should do pairwise-comparison of video streams and discard xvid, mpeg, etc. if H264+ versions exist.

**Comment by alanorth** [comment]
> _preserve dubs from older rips_

Dubs? Yuck! For kids that's cool, but for me as an adult I prefer the original audio with subtitles. :stuck_out_tongue:

> _Ideally, curator merge should do pairwise-comparison of video streams and discard xvid, mpeg, etc. if H264+ versions exist._

That would be cool! Looking forward.

**Comment by AlexAltea** [comment]
> Dubs? Yuck! For kids that's cool, but for me as an adult I prefer the original audio with subtitles. 😛

Oh 100% agreed! Dubs are not for me, but other users of my Jellyfin instance want them. 😄

**Comment by AlexAltea** [comment]
I've managed to implement this, and indeed, it successfully tags the audio tracks in AVI/RIFF files (VLC shows it), but it messes up the metadata which causes some synchronization issues:

![image](https://user-images.githubusercontent.com/5306886/215877999-c87036d1-a4d2-40cf-8d22-37215d8781ab.png)

I'm currently debugging this.

**Comment by AlexAltea** [comment]
Apparently this issue is unavoidable (?), see https://stackoverflow.com/questions/29850246/ffmpeg-avformat-write-header-changing-my-stream-time-base

However, this was already happening with `curator-tag` even before adding support for `IAS1-9` tags, so I've committed the changes, adding simply a warning whenever the input media is an AVI file.

I've tested with my collection, and most audio streams are unaffected by `ffmpeg` changing time bases. But there's been few poorly-made rips that got butchered after tagging. I think this is unavoidable.

I'll work on `curator-sync` next to fix those rips.
